Add support to modify precomputed predicate metadata upon adding/removal of a pod

bsalamat opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**: This PR adds capability to change precomputed predicate metadata and let's us add/remove pods to the precomputed metadata efficiently without the need ot recomputing everything upon addition/removal of pods. This PR is needed as a part of adding preemption logic to the scheduler.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
To make the review process a bit easier, there are three commits. The cleanup commit is only moving code and renaming some functions, without logic changes.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```
ref/ #47604
ref/ #48646

/assign @wojtek-t 

@kubernetes/sig-scheduling-pr-reviews @davidopp

bsalamat committed:
Add support to recompute partial predicate metadata upon adding/removing pods

bsalamat committed:
bazel update

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -17,17 +17,53 @@ limitations under the License.
+	podBestEffort                      bool
+	podRequest                         *schedulercache.Resource
+	podPorts                           map[int]bool
+	matchingAntiAffinityTerms          map[types.UID][]matchingPodAntiAffinityTerm //key is a pod UID with the anti-affinity rule.
```
nit: please avoid such long lines

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -17,17 +17,53 @@ limitations under the License.
+	podRequest                         *schedulercache.Resource
+	podPorts                           map[int]bool
+	matchingAntiAffinityTerms          map[types.UID][]matchingPodAntiAffinityTerm //key is a pod UID with the anti-affinity rule.
+	serviceAffinityInUse               bool
```
Why do we need this bool?
If the two slices below are nulls, it should have the same effect as setting this to false. In that case, I would prefer to remove this to avoid growing this structure infinitely - it's already pretty big.

bsalamat replied:
This bool tells us whether the service affinity predicate is actually enabled. The affinity predicate may be enabled, but the two slices may still be empty because the pod matches no services and no other pods with the same labels exist.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,66 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		deletedPodIndex := -1
+		for i, pod := range meta.serviceAffinityMatchingPodList {
+			if pod.GetUID() == deletedPod.GetUID() {
+				deletedPodIndex = i
```
This is overcomplicated. You simply should move the body of the below if here and get rid of "deletedPodIndex".

bsalamat replied:
Done. I was trying to avoid intervening with the for loop iterator, but it doesn't matter as we are breaking right after removing the element.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,66 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+
+// AddPod changes predicateMetadata assuming that `newPod` is added to the
+// system.
+func (meta *predicateMetadata) AddPod(addedPod *v1.Pod, nodeInfo *schedulercache.NodeInfo) error {
```
Do we really need this method?

My personal feeling is that RemovePod is the only think you need to implement preemption. Why do you want to add pods?

bsalamat replied:
We do need this. The preemption first removes all the lower priority pods and after adding the pending pod, tries to reprieve as many removed pods as possible. This method is useful when adding the pending pod and adding those removed pods back.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,66 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	// Add matching anti-affinity terms of the addedPod to the map.
+	if podMatchingTerms, err := getMatchingAntiAffinityTermsOfExistingPod(meta.pod, addedPod, nodeInfo.Node()); err == nil {
+		if len(podMatchingTerms) > 0 {
+			meta.matchingAntiAffinityTerms[addedPodUID] = append(meta.matchingAntiAffinityTerms[addedPodUID], podMatchingTerms...)
```
This shouldn't be append in my opinion - there should be exactly one pod with a given UID - if there is already something for the uid, this seems like a bug.

bsalamat replied:
There is going to be only one pod, but that one pod may have a few matching anti-affinity terms.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,352 @@
+			return false
+		}
+	}
+	return true
```
This only works for lists without duplicated. It will return true for [1, 1, 2] and [1, 2, 2] (which may or may not be a problem in that case, but definitely deserves a comment).

bsalamat replied:
Deleted the function. It was unused.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,352 @@
+// predicateMetadataEquivalent returns true if the two metadata are equivalent.
+// Note: this function does not compare podRequest.
+func predicateMetadataEquivalent(meta1, meta2 *predicateMetadata) error {
+	if meta1.pod != meta2.pod {
```
Should we really compare pointers to pods here?

wojtek-t replied:
Shouldn't it be DeepEqual?

bsalamat replied:
In this test environment where we pass the same pod pointer to the metadata producers, this is fine. I changed it to DeepEqual anyway to avoid confusion.

wojtek-t reviewed (commented):
I skimmed through the PR and it seems reasonable to me.
I will try to review it in the upcoming days.

bsalamat reviewed (commented):
Thanks a lot @wojtek-t for your prompt review.

bsalamat commented:
@wojtek-t Other than addressing the comments, I also changed the key of the anti-affinity terms from pod UID to pod fullname, in a separate commit. Pod UID was not consistent with the rest of the codebase and made writing tests a bit more difficult.
Could you please take another look?

bsalamat commented:
/retest

bsalamat commented:
ping @wojtek-t. Could you please take another look at your earliest convenience. All of my other PRs to add preemption depend on this one.

wojtek-t commented:
Yes, sorry. Will take a look into it today.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -17,17 +17,53 @@ limitations under the License.
+	podPorts      map[int]bool
+	//key is a pod full name with the anti-affinity rules.
+	matchingAntiAffinityTerms          map[string][]matchingPodAntiAffinityTerm
+	serviceAffinityInUse               bool
```
> This bool tells us whether the service affinity predicate is actually enabled. The affinity predicate may be enabled, but the two slices may still be empty because the pod matches no services and no other pods with the same labels exist.

Correct. But in that case, I think that if both slices are empty no matter if this bool is true or false, the results will be exactly the same.
So basically, what I'm saying is that removing this doesn't change anything and will make the code a bit easier :)

wojtek-t replied:
OK - the argument with AddPod is reasonable. 

SGTM

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,63 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	// So, if the namespace of the first one is not the same as the namespace of the
+	// deletedPod, we don't need to check the list, as deletedPod isn't in the list.
+	if meta.serviceAffinityInUse &&
+		len(meta.serviceAffinityMatchingPodList) > 0 &&
```
This condition is not needed - if the length is zero, the for inside if will simply be no-op.
Let's remove it.

wojtek-t replied:
[In Go you can safely iterate over nil slices].

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,63 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		deletedPod.Namespace == meta.serviceAffinityMatchingPodList[0].Namespace {
+		for i, pod := range meta.serviceAffinityMatchingPodList {
+			if schedutil.GetPodFullName(pod) == deletedPodFullName {
+				meta.serviceAffinityMatchingPodList = append(meta.serviceAffinityMatchingPodList[:i], meta.serviceAffinityMatchingPodList[i+1:]...)
```
nit: please split this line - it's too long now.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,63 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	// Add matching anti-affinity terms of the addedPod to the map.
+	if podMatchingTerms, err := getMatchingAntiAffinityTermsOfExistingPod(meta.pod, addedPod, nodeInfo.Node()); err == nil {
+		if len(podMatchingTerms) > 0 {
+			meta.matchingAntiAffinityTerms[addedPodFullName] = append(meta.matchingAntiAffinityTerms[addedPodFullName], podMatchingTerms...)
```
nit: please split this line

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,63 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	}
+	// Add matching anti-affinity terms of the addedPod to the map.
+	if podMatchingTerms, err := getMatchingAntiAffinityTermsOfExistingPod(meta.pod, addedPod, nodeInfo.Node()); err == nil {
+		if len(podMatchingTerms) > 0 {
```
Please remove this if - it's not neded, you can do it unconditionally:
```
package main

import (
	"fmt"
)

func main() {
	a := []string{"aa"}
	var b []string
	c := []string{}
	d := []string{"ee"}
	
	x := append(a, b...)
	x = append(x, c...)
	x = append(x, d...)
	fmt.Printf("%#v", x)
}
```

bsalamat replied:
When running the tests, the entry may not exist in the map and that's why I had added that initial check. I changed the code to make it clear.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,63 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	// of matching pods if applicable.
+	if meta.serviceAffinityInUse && addedPod.Namespace == meta.pod.Namespace {
+		selector := CreateSelectorFromLabels(meta.pod.Labels)
+		if selector.Matches(labels.Set(addedPod.Labels)) {
```
I don't understand it. Are we checking if selector created from pod labels is maching a newly added pod? This seems incorrect to me...

wojtek-t replied:
I mean, serviceAffinity is about services - there should be some service-related computation here...

wojtek-t replied:
Yeah - sorry.

This predicate surprises me every time I'm seeing it....

wojtek-t replied:
In fact, I think we should get rid of this predicate. I'ts  unfortunate that we can't do it due to backward compatibility reasons...

bsalamat replied:
Yes. Luckily it is not enabled by default.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,332 @@
+	if len(meta1.podPorts) != len(meta2.podPorts) {
+		return fmt.Errorf("podPorts are not equal.")
+	}
+	for k1, v1 := range meta1.podPorts {
```
Shouldn't DeepEqual work fine for this?

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,332 @@
+			return fmt.Errorf("podPorts are not equal.")
+		}
+	}
+	for k1, v1 := range meta1.matchingAntiAffinityTerms {
```
This is test - we don't need to be super efficient here - simpler code is more important here.

So instead of that, we can compare two slices by:
- sorting both of them
- and then calling DeepEqual

[Yes - I know you need to write comparator for sorting, but this method will be so much simpler then].

The same below for service-afffinity things.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,332 @@
+		allPodLister := schedulertesting.FakePodLister(append(test.existingPods, test.addedPod))
+		// getMeta creates predicate meta data given the list of pods.
+		getMeta := func(lister schedulertesting.FakePodLister) (*predicateMetadata, map[string]*schedulercache.NodeInfo) {
+			nodeInfoMap := map[string]*schedulercache.NodeInfo{}
```
Please split this into a separate utility function (in this file).

wojtek-t replied:
I mean splitting the part computing nodeInfoMap.

I think that we used to have such function in the past, but I can't find it now...

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1105,7 +1077,8 @@ func getMatchingAntiAffinityTerms(pod *v1.Pod, nodeInfoMap map[string]*scheduler
 				if priorityutil.PodMatchesTermsNamespaceAndSelector(pod, namespaces, selector) {
-					nodeResult = append(nodeResult, matchingPodAntiAffinityTerm{term: &term, node: node})
+					existingPodFullName := schedutil.GetPodFullName(existingPod)
+					nodeResult[existingPodFullName] = append(nodeResult[existingPodFullName], matchingPodAntiAffinityTerm{term: &term, node: node})
```
nit: please split the line

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1117,56 +1090,73 @@ func getMatchingAntiAffinityTerms(pod *v1.Pod, nodeInfoMap map[string]*scheduler
 		matchingTerms = predicateMeta.matchingAntiAffinityTerms
 	} else {
-		allPods, err := c.podLister.List(labels.Everything())
+		filteredPods, err := c.podLister.FilteredList(nodeInfo, labels.Everything())
```
those actually are "allPods" - labels.Everything won't filter anything, right?

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -106,6 +106,29 @@ func (cache *schedulerCache) List(selector labels.Selector) ([]*v1.Pod, error) {
+// PodFilter is a simple interface that helps filter out pods.
+// This is here only to avoid circular dependency between schedulercache and
+// algorithm. Ideally, it should have been in algorithm/types.go.
+type PodFilter interface {
```
I'm not a fan of this approach. I would prefer defining:
```
type func(*v1.Pod) bool PodFilter
```

and passing explicitly nodeInfo.Filter() to FilteredList.

Or do you have any strong reason of not doing that?

bsalamat replied:
No strong reason. Done as you suggested.

wojtek-t reviewed (commented):
I added some comments.

Also pleas squash commits - I'm reviewing the whole PR anyway, because I never remember what I did before :P

bsalamat reviewed (commented):
Thanks a lot, @wojtek-t!
Please take another look.

bsalamat commented:
/retest

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,70 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		if found {
+			existingTerms = append(existingTerms, podMatchingTerms...)
+		} else {
+			if len(podMatchingTerms) > 0 {
```
This should be checked in the outer if - there is no point in doing all those operations if this is empty.

```
if len(podMatchingTerms) > 0 {
  if existingTerms, found := meta.matchingAntiAffinityTerms[addedPodFullName]; found {
    existingTerms = append(existingTerms, podMatchingTerms...)
  } else {
    meta.matchingAntiAffinityTerms[addedPodFullName] = podMatchingTerms
  }
}
```

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,70 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		return fmt.Errorf("Invalid node in nodeInfo.")
+	}
+	// Add matching anti-affinity terms of the addedPod to the map.
+	if podMatchingTerms, err := getMatchingAntiAffinityTermsOfExistingPod(meta.pod, addedPod, nodeInfo.Node()); err == nil {
```
would prefer to limit the amount of nested ifs. So it would be easier to follow (and more standard) to do:

podMatchingTerms, err := getMatchingAntiAffinityTermsOfExistingPod(meta.pod, addedPod, nodeInfo.Node());
if err != nil {
  return err
}
// The rest of code.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,324 @@
+	for !reflect.DeepEqual(meta1.podPorts, meta2.podPorts) {
+		return fmt.Errorf("podPorts are not equal.")
+	}
+	for k1, v1 := range meta1.matchingAntiAffinityTerms {
```
> It might be, but sort interface needs three functions. It should be done for the service affinity as well. We should also be careful that the comparison function almost never generates equal cases, as the order of equal elements in the expected list may not necessarily be the same as what the tests generate and could cause false positives. Given all this, probably sorting and doing DeepEqual may need even more code.

I agree that sort needs 3 function. But those are super trivial function and well separated.

And yes - I agree that in total this might be a bit more code, but it will be better structured and easier to follow.
This function now has 70 lines of code and it's hard to follow it.
And what if we add few more fields to predicateMetadata? We may end up with comparison function having hundreds of line.

If you really don't want to do the sorting (though i think it's easier and less error prone), at the very least please extract comparing those slices to indviidual functions with a TODO to simplify them.

bsalamat replied:
Alright. Done. As expected there is more code now, but I agree that it is probably easier to follow.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1117,56 +1092,73 @@ func getMatchingAntiAffinityTerms(pod *v1.Pod, nodeInfoMap map[string]*scheduler
+func (c *PodAffinityChecker) getMatchingAntiAffinityTerms(pod *v1.Pod, allPods []*v1.Pod) (map[string][]matchingPodAntiAffinityTerm, error) {
+	result := make(map[string][]matchingPodAntiAffinityTerm)
+	for _, existingPod := range allPods {
+		existingPodNode, err := c.info.GetNodeInfo(existingPod.Spec.NodeName)
```
This may affect performance, because with this change you're computing it for every pod, and before we were doing it only for pods with affinity/anti-affinity.

I think we should change it back to how it was.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1117,56 +1092,73 @@ func getMatchingAntiAffinityTerms(pod *v1.Pod, nodeInfoMap map[string]*scheduler
+		if err != nil {
+			return nil, err
+		}
+		existingPodFullName := schedutil.GetPodFullName(existingPod)
```
if len(existingPodMatchingTerms) > 0 {
...
}

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1117,56 +1092,73 @@ func getMatchingAntiAffinityTerms(pod *v1.Pod, nodeInfoMap map[string]*scheduler
-				podName(pod), node.Name, term.term)
-			return false
+	for _, terms := range matchingTerms {
+		for _, term := range terms {
```
This will result in creating a copy of each of term.

Instead you should do:
```
for i := range terms {
  term := &terms[i]
   ...
```

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -106,6 +106,20 @@ func (cache *schedulerCache) List(selector labels.Selector) ([]*v1.Pod, error) {
 	return pods, nil
 }
 
+func (cache *schedulerCache) FilteredList(podFilter PodFilter, selector labels.Selector) ([]*v1.Pod, error) {
```
Instead of duplicating the code please do:

```
func (cache *schedulerCache) List(selector labels.Selector) ([]*v1.Pod, error) {
  return cache.filteredList(selector, alwaysTrue)
}

func (cache *schedulerCache) FilteredList(podFilter PodFilter, selector labels.Selector) ([]*v1.Pod, error) {
  return cache.filteredList(selector, podFilter)
}

func (cache *schedulerCache) filteredList(podFilter PodFilter, selector labels.Selector) ([]*v1.Pod, error) {
  // as you FilterList is now
}
```

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1117,56 +1092,73 @@ func getMatchingAntiAffinityTerms(pod *v1.Pod, nodeInfoMap map[string]*scheduler
 		matchingTerms = predicateMeta.matchingAntiAffinityTerms
 	} else {
-		allPods, err := c.podLister.List(labels.Everything())
+		filteredPods, err := c.podLister.FilteredList(nodeInfo.Filter, labels.Everything())
```
Wait - this is completely changing the logic.

With nodeInfo.Filter you will only get pods from a given node, not all pods. That is not correct in my opinion.

wojtek-t reviewed (commented):
I added another portion of comments. Most of them are relativley minor, there is no bigger one.

bsalamat reviewed (commented):
Thanks, @wojtek-t! PTAL.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,70 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	if len(podMatchingTerms) > 0 {
+		existingTerms, found := meta.matchingAntiAffinityTerms[addedPodFullName]
+		if found {
+			meta.matchingAntiAffinityTerms[addedPodFullName] = append(existingTerms, podMatchingTerms...)
```
nit: please split the line

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -52,9 +88,70 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	if meta.serviceAffinityInUse && addedPod.Namespace == meta.pod.Namespace {
+		selector := CreateSelectorFromLabels(meta.pod.Labels)
+		if selector.Matches(labels.Set(addedPod.Labels)) {
+			meta.serviceAffinityMatchingPodList = append(meta.serviceAffinityMatchingPodList, addedPod)
```
nit: please split the line

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,363 @@
+)
+
+// sortableAntiAffinityTerms lets us to sort anti-affinity terms.
+type sortableAntiAffinityTerms struct {
```
You can simply define this as:
```
type sortableAntiAffinityTerms []matchinPodAntiAffinityTerm
```

Will be simpler. Same for pods below.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,363 @@
+}
+
+func (s *sortablePods) Less(i, j int) bool {
+	if s.Items[i].Name < s.Items[j].Name {
```
This isn't necessarily good sorting function.
You can have two pods with the same name but in different namespaces. And those aren't equal.

Should be:
```
return s[i].Namespace < s[j].Namespace || (s[i].Namespace == s[j].Namespace && s[i].Name < s[j].Name)
```

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,363 @@
+}
+
+func (s *sortableServices) Less(i, j int) bool {
+	if s.Items[i].Name < s.Items[j].Name {
```
Same here.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,363 @@
+func (s *sortableAntiAffinityTerms) Less(i, j int) bool {
+	t1, t2 := s.Items[i], s.Items[j]
+	return t1.node.Name < t2.node.Name ||
+			(t1.node.Name == t2.node.Name && len(t1.term.Namespaces) < len(t2.term.Namespaces)) ||
```
This is a bit unfortunate.
To make it a bit more readable, we can try changing to:
```
if t1.node.Name != t2.node.Name {
  return t1.node.Name < t2.node.Name
}
// and now compare the rest similarly
```

It's effectively the same, but simpler to follow.

bsalamat replied:
Done

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,363 @@
+	for !reflect.DeepEqual(meta1.podPorts, meta2.podPorts) {
+		return fmt.Errorf("podPorts are not equal.")
+	}
+	sortAntiAffinityTerms(meta1.matchingAntiAffinityTerms)
```
Thanks - it's much easier to follow now!

wojtek-t reviewed (commented):
A couple minor comments. I will most probably LGTM once applied. Thanks for bothering with me!

bsalamat reviewed (commented):
Thanks, @wojtek-t! PTAL.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -0,0 +1,360 @@
+	if t1.node.Name != t2.node.Name {
+		return t1.node.Name < t2.node.Name
+	}
+	if t1.node.Name == t2.node.Name && len(t1.term.Namespaces) != len(t2.term.Namespaces) {
```
You don't need to check if node.Names are equal. They are, otherwise we would return in above if.

Same for everything below.

bsalamat replied:
Done

wojtek-t reviewed (commented):
The last minor nit.

Please squash the last commit and I will approve.

bsalamat commented:
Thanks a lot @wojtek-t for taking a lot of time to review this PR.

davidopp commented:
Thanks a ton for reviewing this @wojtek-t

wojtek-t commented:
/lgtm
